src/main/java/com/google/devtools/build/lib/remote/RemoteDownloaderFactory.java

```diff
@@ -0,0 +1,43 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.remote;
+
+import com.google.common.base.Preconditions;
+import com.google.devtools.build.lib.bazel.repository.downloader.Downloader;
+import com.google.devtools.build.lib.bazel.repository.downloader.DownloaderFactory;
+import java.util.Optional;
+
+public final class RemoteDownloaderFactory implements DownloaderFactory {
+
+  private volatile Downloader remoteDownloader;
+
+  public void init(Downloader remoteDownloader) {
+    Preconditions.checkState(this.remoteDownloader == null);
+    this.remoteDownloader = remoteDownloader;
+  }
+
+  public void reset() {
+    this.remoteDownloader = null;
+  }
+
+  @Override
+  public Optional<Downloader> create() {
```

Comment:
This implementation technically violates the specification on the interface, which says:
Returns a ***new*** Downloader, or empty.

Reply:
Changed `DownloaderFactory::create()` to not say "new".